File: 1544-count-good-nodes-in-binary-tree/count-good-nodes-in-binary-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function goodNodes(root: TreeNode | null): number {
    const dfs = (node: TreeNode | null, maxSoFar: number): number => {
        if (node === null) return 0

        let good = node.val >= maxSoFar ? 1 : 0

        maxSoFar = Math.max(maxSoFar, node.val)
        
        good += dfs(node.left, maxSoFar)
        good += dfs(node.right, maxSoFar)

        return good
    }

    return dfs(root, root.val)
};

File: 1544-count-good-nodes-in-binary-tree/count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(node, max_):
            if not node:
                return 0

            count = 0
            if node.val >= max_:
                count += 1
                max_ = node.val

            left = dfs(node.left, max_)
            right = dfs(node.right, max_)

            return left + right + count
        return dfs(root, float('-inf'))